### PR TITLE
redirect user to payment when the card doesn't have enough credit

### LINKED ISSUE #9660: [Split Checkout] Failed StripeSCA payments redirect to the Details step
## Description

If a payment fails - for example, due to insufficient funds - then the customer is redirected to the `Details` step. This forces the user to re-select a shipping method.

## Expected Behavior

If a payment fails - for example, due to insufficient funds - then the customer should be redirected to the `Payment` step.

## Actual Behaviour

If a payment fails - for example, due to insufficient funds - then the customer is redirected to the `Details` step.

## Steps to Reproduce

1. Test checking out with a the Stripe card x-3178
2. Grant authorization, when prompted to.
3. Notice the payment fails (`insufficient_funds`), but notice that you're redirected into the Details step

## Animated Gif/Screenshot

image

### PR DESCRIPTION
**Pull request by abdellani**
#### What? Why?

- Closes #9660

#### What should we test?

- described on: https://github.com/openfoodfoundation/openfoodnetwork/issues/9660

#### Release notes

Changelog Category: User facing changes | Technical changes

The title of the pull request will be included in the release notes.

### CODE REVIEW

File: app/controllers/concerns/order_completion.rb [hunk 1]
```diff
@@ -87,7 +87,6 @@ def processing_failed(error = RuntimeError.new(order_processing_error))
       payload.add_metadata :order, @order
     end
     flash[:error] = order_processing_error if flash.blank?
-    Checkout::PostCheckoutActions.new(@order).failure
```
Comment (mkllnk) [inline app/controllers/concerns/order_completion.rb:90]: I think that this line is quite important in other cases. For example, if something goes out of stock or the order cycle closes then we need to reset some data. I can't remember the exact circumstances but we had bugs without it. This is actually the only line calling the `failure` method. So maybe you need two methods:

* `processing_failed` and
* `payment_failed`

`payment_failed` would be the Bugsnag notification and flash error while `processing_failed` also calls the `failure` method.

**Review by mkllnk — COMMENTED** [review]
Great work. I left a comment but I didn't investigate it. I could be totally wrong. So I don't know if it needs changing or not. It would be good to find the original issue which was the reason to add the `failure` method and see if the bug re-appears with your version or not.

File: app/controllers/concerns/order_completion.rb (hunk 1 above)
Comment (dacook) [reply app/controllers/concerns/order_completion.rb:90]: It seems like we want to treat "processing failures" consistently with `processing_failed`, but not all are to be treated with `PostCheckoutActions#failure`. If it is important, I would have expected that to be apparent in the controller or concern specs. So maybe they need updating too!

Maybe that method should be renamed more specifically about what failures it's for (I've not looked into what it's for yet!).
And so as Maikel suggests, we only call it if it's not a payment error.

**Review by dacook — COMMENTED** [review]
The redirect changes look good, but we'd better treat the PostCheckoutActions more carefully as Maikel mentions.

File: app/controllers/concerns/order_completion.rb (hunk 1 above)
Comment (abdellani) [reply app/controllers/concerns/order_completion.rb:90]: Ok, I extracted that call to `processing_failed`

**Review by dacook — APPROVED** [review]
👍 This addresses Maikel's comment and I think makes the bug fix safe.

 I have to admit I'm not familiar with the subject matter so am assuming that the results of `PostCheckoutOptions#failure` don't require further refactoring for this case.

**Review by mkllnk — APPROVED** [review]
Nice!

**Comment by filipefurtad0** [comment]
Hey @abdellani ,

Awesome :tada: 
After this PR we get the correct behavior:

With `x-3178` 3D card, 

image

With `x-9995` non-3D card, 

image

In both cases, the user gets redirected to the `/payment` step.


With a valid 3D card `x-3155`, the payment is accepted after authorization, and the user is redirected to the order confirmation page.

Merging :rocket:

**Pull request merged**